### ISSUE TITLE
Regression in USER-OMP: allocations too small

### ISSUE THREAD

**Issue by rbberger** [lammps/lammps#2081]
**Summary**

The new AtomVec implementation creates a regression with the USER-OMP package. It assumes `comm->nthreads` does not change. In some cases (e.g. no atom style is set), `FixOMP` sets `comm->nthreads` after `AtomVec` is created.

```
$ lmp -in in.balance.kspace -sf omp -pk omp 4
LAMMPS (5 May 2020)
  using 1 OpenMP thread(s) per MPI task
using multi-threaded neighbor list subroutines
set 4 OpenMP thread(s) per MPI task
...
```
In this case, AtomVec will set its local `nthreads` to 1 and FixOMP will change `comm->nthreads` to 4. All `AtomVec::grow` executions will continue with `nthreads == 1`. This eventually leads to a segfault when accessing the force array.

**LAMMPS Version and Platform**

current master

**Expected Behavior**

Memory buffers get allocated for `comm->nthreads == 4`

**Actual Behavior**

Segmentation fault, since allocation happens for `AtomVec::nthread == 1`

**Steps to Reproduce**

```bash
mkdir build
cd build
cmake -D PKG_KSPACE=on -D PKG_USER-OMP=on ../cmake
make
LAMMPS_BINARY=$PWD/lmp
cd ../examples/balance/
$LAMMPS_BINARY -in in.balance.kspace -sf omp -pk omp 4
```

**Workaround**

1. Add `atom_style` to input script
OR
2. Change all mentions of `nthreads` in atom_vec.cpp to `comm->nthreads`

**Comment by akohlmey** [comment]
This is something that resonates with a habit in programming LAMMPS classes that has been bothering me for a long time: there are too many variables defined inside LAMMPS classes that need not be there because they are redundant or are only used locally in a single function. `nthreads` is just one of them, there are others (like `me` or `natoms` for example).

And that touches some other issue: some of these variables have side effects when changed, i.e. they should not be changed without changing other, dependent variables, and that burden should not be on the person programming some code that changes a property, but rather should be on the class that maintains that data. In other words, one should not be allowed to directly change this variable, but there should be a "setter" method to modify this property (and its derived properties) and correspondingly a "getter" method to get properties that are currently stored in a redundant fashion, e.g. instead of having "xperiodic" in domain.h, there should be get_xperiodic() which returns periodic[0]. Consequently, that also means that any code that uses this data should be assuming it can change, which is more obvious with a "getter" method than when using a class member variable, especially to people who want to extend some class and have not written the original implementation.

**Comment by akohlmey** [comment]
After looking at the code, the situation is a bit more complex, as there is also the "threads" array with entries that are either 1 or the number of threads.

So the solution has to be expanded beyond using `comm->nthreads` instead of `nthreads`.
I think the most straightforward solution would be to make `threads` an array of type `bool`
and then have locally `int nthreads = comm->nthreads` and instead of `threads[i]` use `threads[i] ? nthreads : 1`.

The alternative would be to have a function `update_threads()` that would detect a change between `nthreads` and `comm->nthreads` and update the `threads` array accordingly, but that may be complicated to do when `nthreads` was previously set to 1.

**Comment by akohlmey** [comment]
The first option is now implemented in PR #2080
